### PR TITLE
cmake: add SDL_ICONV option to enable/disable iconv

### LINKED ISSUE #6334: SDL_iconv forbid use of libiconv even if installed
Hi, is it possible to disable libiconv inclusion even if the library is installed?

I compile my project in MSYS2 and libiconv is installed for some other dependency. However, I plan to run it outside MSYS2 and I am not keen to distribute libiconv.dll alongside it.

SDL provides a fallback but that is only enabled if it cannot find libiconv from the system (which I can't just uninstall). Is there any way around this? For example adding a CMake flag or something.

### PR DESCRIPTION
**Pull request by madebr**
FIxes #6334

Tested with targeting Linux and MinGW x64 (using wine)

### CODE REVIEW

**Comment by sezero** [comment]
This shouldn't be cmake-only: autotools will need an equivalent, if this gets in.

File: configure.ac [hunk 1]
```diff
@@ -4297,7 +4305,11 @@ dnl BeOS support removed after SDL 2.0.1. Haiku still works.  --ryan.
         SOURCES="$SOURCES $srcdir/src/video/uikit/*.m"
         SUMMARY_video="${SUMMARY_video} uikit"
         have_video=yes
-        EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lm -liconv -lobjc"
+        EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lm"
+        if test x$enable_libc = xyes; then
```
Comment (slouken) [inline configure.ac:4309]: You probably meant $enable_iconv here?

File: configure.ac [hunk 2]
```diff
@@ -4559,7 +4571,7 @@ dnl BeOS support removed after SDL 2.0.1. Haiku still works.  --ryan.
 
         # Set up the core platform files
         SOURCES="$SOURCES $srcdir/src/core/os2/*.c"
-        if test x$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes; then
+        if test x$enable_libc = xyes -a ( thenx$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes ); then
```
Comment (slouken) [inline configure.ac:4574]: and here? Also I think the "then" is a syntax error?

File: configure.ac (hunk 1 above)
Comment (madebr) [reply configure.ac]: Yes yes yes. So sorry!
Fixed in https://github.com/libsdl-org/SDL/commit/5d62b3bdb0a216732b2aae63015e3b2a014ec865

File: configure.ac [hunk 3]
```diff
@@ -4559,7 +4571,7 @@ dnl BeOS support removed after SDL 2.0.1. Haiku still works.  --ryan.
 
         # Set up the core platform files
         SOURCES="$SOURCES $srcdir/src/core/os2/*.c"
-        if test x$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes; then
+        if test x$enable_iconv = xyes -a ( x$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes ); then
             SOURCES="$SOURCES $srcdir/src/core/os2/geniconv/*.c"
         fi
```
Comment (madebr) [inline configure.ac:4576]: @sezero 
Will os2 still build? No missing symbols?

Comment (sezero) [reply configure.ac]: It should - no reason otherwise. (Don't have a way of build testing against klibc now, though.)

**Comment by slouken** [comment]
Looks good, thanks!

**Comment by madebr** [comment]
> Looks good, thanks!

Thanks, but I would have waited with merging until CI has succeeded.
I think I performed a few force pushes too in succession and tripped some overload protection.

**Comment by slouken** [comment]
Sorry, I thought that I saw that it had. Maybe my GitHub view hadn't updated after your force push? In any case, it looks good and we can fix up any CI issues.

**Comment by sezero** [comment]
While you are fixing any possible aftermath:
`Enable character set conversion through iconv` sounds misleading. A
_probably_ better description would be specifying `system` in there,
like `Enable character set conversion through system-installed iconv`
(or something with better wording than mine.)

**Comment by slouken** [comment]
This sounds like a good change. @sezero, feel free to push a commit for this.

**Comment by slouken** [comment]
Maybe the wording could be:
Use system-installed libiconv for character set conversion

**Comment by sezero** [comment]
> Maybe the wording could be: Use system-installed libiconv for character set conversion

iconv & co doesn't have to come from libiconv, it can be in libc.

**Comment by sezero** [comment]
I suggest the following with both description and the configurator option name change:
OK to apply ?

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index adbf9eb..5eadd13 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -448,7 +448,7 @@ set_option(SDL_DIRECTFB            "Use DirectFB video driver" OFF)
 dep_option(SDL_DIRECTFB_SHARED     "Dynamically load directfb support" ON "SDL_DIRECTFB" OFF)
 set_option(SDL_DUMMYVIDEO          "Use dummy video driver" ON)
 dep_option(SDL_IBUS                "Enable IBus support" ON ${UNIX_SYS} OFF)
-set_option(SDL_ICONV               "Support character set conversion through libiconv" ON)
+set_option(SDL_SYSICONV            "Use iconv() from system-installed libraries" ON)
 set_option(SDL_OPENGL              "Include OpenGL support" ON)
 set_option(SDL_OPENGLES            "Include OpenGL ES support" ON)
 set_option(SDL_PTHREADS            "Use POSIX threads for multi-threading" ${SDL_PTHREADS_ENABLED_BY_DEFAULT})
@@ -1024,7 +1024,7 @@ if(SDL_LIBC)
       endif()
     endif()
 
-    if(SDL_ICONV)
+    if(SDL_SYSICONV)
       check_library_exists(iconv iconv_open "" HAVE_LIBICONV)
       if(HAVE_LIBICONV)
         list(APPEND EXTRA_LIBS iconv)
diff --git a/configure.ac b/configure.ac
index 6cd98df..c2b4d73 100644
--- a/configure.ac
+++ b/configure.ac
@@ -319,9 +319,9 @@ AC_ARG_ENABLE(libc,
               , enable_libc=yes)
 
 dnl See whether we are allowed to use libiconv
-AC_ARG_ENABLE(iconv,
-[AS_HELP_STRING([--enable-iconv], [Enable character set conversion through iconv [default=yes]])],
-              , enable_iconv=yes)
+AC_ARG_ENABLE(sysiconv,
+[AS_HELP_STRING([--enable-sysiconv], [Use iconv() from system-installed libraries [default=yes]])],
+              , enable_sysiconv=yes)
 
 if test x$enable_libc = xyes; then
     AC_DEFINE(HAVE_LIBC, 1, [ ])
@@ -353,7 +353,7 @@ dnl Checks for library functions.
     AC_CHECK_LIB(m, pow, [LIBS="$LIBS -lm"; EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lm"])
     AC_CHECK_FUNCS(acos acosf asin asinf atan atanf atan2 atan2f ceil ceilf copysign copysignf cos cosf exp expf fabs fabsf floor floorf trunc truncf fmod fmodf log logf log10 log10f lround lroundf pow powf round roundf scalbn scalbnf sin sinf sqrt sqrtf tan tanf)
 
-    if test x$enable_iconv = xyes; then
+    if test x$enable_sysiconv = xyes; then
         AC_CHECK_LIB(iconv, iconv_open, [LIBS="$LIBS -liconv"; EXTRA_LDFLAGS="$EXTRA_LDFLAGS -liconv"])
         AC_CHECK_FUNCS(iconv)
     fi
@@ -4306,7 +4306,7 @@ dnl BeOS support removed after SDL 2.0.1. Haiku still works.  --ryan.
         SUMMARY_video="${SUMMARY_video} uikit"
         have_video=yes
         EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lm"
-        if test x$enable_iconv = xyes; then
+        if test x$enable_sysiconv = xyes; then
             EXTRA_LDFLAGS="$EXTRA_LDFLAGS -liconv"
         fi
         EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lobjc"
@@ -4571,7 +4571,7 @@ dnl BeOS support removed after SDL 2.0.1. Haiku still works.  --ryan.
 
         # Set up the core platform files
         SOURCES="$SOURCES $srcdir/src/core/os2/*.c"
-        if test x$enable_iconv = xyes; then
+        if test x$enable_sysiconv = xyes; then
             if test x$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes; then
                 SOURCES="$SOURCES $srcdir/src/core/os2/geniconv/*.c"
             fi
diff --git a/configure b/configure
index 3e0cadf..dbaa3fa 100755
--- a/configure
+++ b/configure
@@ -840,7 +840,7 @@ enable_largefile
 enable_assertions
 enable_dependency_tracking
 enable_libc
-enable_iconv
+enable_sysiconv
 enable_gcc_atomics
 enable_atomic
 enable_audio
@@ -1632,7 +1632,7 @@ Optional Features:
   --enable-dependency-tracking
                           Use gcc -MMD -MT dependency tracking [default=yes]
   --enable-libc           Use the system C library [default=yes]
-  --enable-iconv          Enable character set conversion through iconv
+  --enable-sysiconv       Use iconv() from system-installed libraries
                           [default=yes]
   --enable-gcc-atomics    Use gcc builtin atomics [default=yes]
   --enable-atomic         Enable the atomic operations subsystem [default=yes]
@@ -18680,12 +18680,12 @@ else $as_nop
 fi
 
 
-# Check whether --enable-iconv was given.
-if test ${enable_iconv+y}
+# Check whether --enable-sysiconv was given.
+if test ${enable_sysiconv+y}
 then :
-  enableval=$enable_iconv;
+  enableval=$enable_sysiconv;
 else $as_nop
-  enable_iconv=yes
+  enable_sysiconv=yes
 fi
 
 
@@ -19896,7 +19896,7 @@ then :
 fi
 
 
-    if test x$enable_iconv = xyes; then
+    if test x$enable_sysiconv = xyes; then
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for iconv_open in -liconv" >&5
 printf %s "checking for iconv_open in -liconv... " >&6; }
 if test ${ac_cv_lib_iconv_iconv_open+y}
@@ -28983,7 +28983,7 @@ printf "%s\n" "#define SDL_VIDEO_RENDER_OGL_ES2 1" >>confdefs.h
         SUMMARY_video="${SUMMARY_video} uikit"
         have_video=yes
         EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lm"
-        if test x$enable_iconv = xyes; then
+        if test x$enable_sysiconv = xyes; then
             EXTRA_LDFLAGS="$EXTRA_LDFLAGS -liconv"
         fi
         EXTRA_LDFLAGS="$EXTRA_LDFLAGS -lobjc"
@@ -29282,7 +29282,7 @@ printf "%s\n" "#define SDL_TIMER_UNIX 1" >>confdefs.h
 
         # Set up the core platform files
         SOURCES="$SOURCES $srcdir/src/core/os2/*.c"
-        if test x$enable_iconv = xyes; then
+        if test x$enable_sysiconv = xyes; then
             if test x$ac_cv_func_iconv != xyes -o x$ac_cv_header_iconv_h != xyes; then
                 SOURCES="$SOURCES $srcdir/src/core/os2/geniconv/*.c"
             fi
```

**Comment by slouken** [comment]
Changing the wording is fine, please leave the option naming the way it was (iconv instead of sysiconv)

**Comment by sezero** [comment]
Still don't like the current option name but I did as you asked and
pushed 294ccba0a23b37fffef62189423444f93732e565.

**Comment by slouken** [comment]
Ah, because you're thinking that `SDL_ICONV` indicates whether SDL is built with iconv support at all, vs pulling it in from system libraries, is that right?

**Comment by sezero** [comment]
Yes

**Comment by slouken** [comment]
From that perspective it makes sense. Do we have any other similar options for C library functions?

**Comment by sezero** [comment]
Doesn't seem so, no. And this particular one is added to deal with a pretty specific case.

**Comment by slouken** [comment]
Okay, let's be verbose about it then, and call it SDL_SYSTEM_ICONV and --enable-system-iconv

**Comment by madebr** [comment]
For cmake, make sure to set `HAVE_SDL_SYSTEM_ICONV` as well.

**Comment by sezero** [comment]
> Okay, let's be verbose about it then, and call it SDL_SYSTEM_ICONV and --enable-system-iconv

OK, I have the patch. The cmake final report issue @madebr mentioned is there, though:
`--   SDL_SYSTEM_ICONV            (Wanted: ON): OFF`

How do I fix that?

**Comment by sezero** [comment]
> How do I fix that?

OK saw your `HAVE_SDL_SYSTEM_ICONV` late. Doing it now and will push shortly.

**Comment by madebr** [comment]
It might also be `HAVE_SYSTEM_ICONV` that you must set. I'm not 100% sure about that.

**Comment by sezero** [comment]
Looks like it's `HAVE_SYSTEM_ICONV`

Question for this final report cosmetics: Do we set `HAVE_SYSTEM_ICONV`
always if `SDL_SYSTEM_ICONV` is ON, or should we also consider either of
`HAVE_LIBICONV` or `HAVE_BUILTIN_ICONV` succeeded? (Or even `HAVE_LIBC`
for that matter.)

**Comment by madebr** [comment]
In my opinion, `libiconv.so` and `libc.so` are both external/system libraries. So if one of both is used, it should be considered ON.

**Comment by madebr** [comment]
The original issue this pr was trying to fix was not to avoid the system library though, but to avoid a dependency on `libiconv.so`.

**Comment by sezero** [comment]
To clarify:

Is it the following first one?
```diff
-    if(SDL_ICONV)
+    if(SDL_SYSTEM_ICONV)
+      set(HAVE_SYSTEM_ICONV TRUE)
```

... or the following second one?
```diff
       check_library_exists(iconv iconv_open "" HAVE_LIBICONV)
       if(HAVE_LIBICONV)
         list(APPEND EXTRA_LIBS iconv)
         set(HAVE_ICONV 1)
+        set(HAVE_SYSTEM_ICONV TRUE)
       else()
         check_library_exists(c iconv_open "" HAVE_BUILTIN_ICONV)
         if(HAVE_BUILTIN_ICONV)
           set(HAVE_ICONV 1)
+          set(HAVE_SYSTEM_ICONV TRUE)
         endif()
       endif()
```

**Comment by madebr** [comment]
The 2nd one. The `HAVE_` variables show whether the option has actually been enable.

`SDL_SYSTEM_ICONV` says: "I want system iconv"
`HAVE_SYSTEM_ICONV` says: "we (don't) have system iconv"


(Personally, I find the use of these variables not very robust)

**Comment by sezero** [comment]
Done

**Pull request merged**